Exercises/1.h

```diff
@@ -0,0 +1,96 @@
+﻿#include <iostream>
+#include <vector>   
+#include <random>
+#include <algorithm>
+#include <time.h>
+#include <boost/graph/graph_traits.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/depth_first_search.hpp>
+
+class disjoint_path_finder {
+
+private:
+
+    typedef boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS> Graph;
+    typedef boost::graph_traits<Graph>::vertex_descriptor vertex_t;
+    typedef boost::graph_traits<Graph>::edge_descriptor edge_t;
+
+    std::vector<int> _color;
+    std::vector<std::pair<int, int>> _pairs;
+    Graph _g;
+    
+    vertex_t get_vertex_by_number(int i, Graph& g) {
+        /*
+        Transform number of vertex |i| in Graph |g| into vertex_t 
+        @param Order number of vertex in the Graph |g|
+        @return vertex_t in the Graph |g|
+        */
+        return (*(vertices(g).first + i));
+    }
+
+    bool my_dfs(vertex_t start, vertex_t end, int c, Graph& g) { 
+/*
+        Finds a path from |start| to |end| in |g| while enumerating the neighbors of each vertex in an order chosen at random. 
+        Marks all the nodes along the found path in the |this->_color| vector with |c|.
+
+        @param start is the start node of the potential path
+        @param end is the end node of the potential path
+        @param c is the _color to mark the nodes in the path with
+        @param g the graph to make the traversal
+        @return true if a path was found and false otherwise
+*/
+        _color[start] = c;
+        if (start == end) return true;
+        std::vector<int> go(degree(start, g));
+        for (int i = 0; i < go.size(); ++i) {
+            go[i] = i;
+        }
+        random_shuffle(go.begin(), go.end());
+        for (int i : go) {
+            auto u = *(adjacent_vertices(start, g).first + i);
+            if ((_color[u] == -1 || _color[u] == c) && my_dfs(u, end, c, g)) return true;
+        }
+        _color[start] = -1;
+        return false;
+    }
+
+public:
+
+    disjoint_path_finder(int n, std::vector<std::pair<int, int>> edges, std::vector<std::pair<int, int>> pairs) : _pairs(pairs) {
+        /*
+        Constructor generates Graph |_g| with |n| vertices and edges from |edges|
+        @param |n| amount of verices
+        @param |edges| - edges of the Graph |_g|
+        @param |_pairs| - pairs of vertices that would form disjoint path decomposition 
+        */
+        srand(time(0));
```

Comment:
Should not be here. There should be a way to control the randomness from the outside (say, in unit tests) to make possible issues reproducible.